README.md
# sca-codeinsight-reports-claim-evidence

The `sca-codeinsight-reports-claim-files` repository is a example custom report for Revenera's Code Insight product. This report allows a user to determine which files within a project contain only evidence that is claimable based on a string comparison to the follow evidence types:

- copyright
- emails/urls
 
Action can be taken while running the report as well. If this value is *True* then the scanned files that contain only the evidence from the specified claimable values will be marked as reviewed and associated to the inventory item provided.

This repository utilizes the following via CDN for the creation of the report artifacts.

- [Bootstrap](https://getbootstrap.com/)
- [DataTables](https://datatables.net/)

## Prerequisites

**Code Insight Release Requirements**

|Repository Tag | Minimum Code Insight Release |
|--|--|
|1.0.x | 2021R2 |

**Repository Cloning**

This repository should be cloned directly into the **$CODEINSIGHT_INSTALLDIR/custom_report_scripts** directory. If no prior custom reports has been installed, this directory may need to be created prior to cloning.

**Submodule Repositories**

This repository contains two submodule repositories for code that is used across multiple projects.  There are two options for cloning this repository and ensuring that the required submodules are also installed.

Clone the report repository use the recursive option to automatically pull in the required submodules

	git clone --recursive

 Alternatively clone the report repository and then clone the submodules separately by entering the cloned directory and then pulling down the necessary submodules code via   

	git submodule init

	git submodule update

**Python Requirements**

The required python modules can be installed with the use of the [requirements.txt](requirements.txt) file which can be loaded via.

    pip install -r requirements.txt    

## Configuration and Report Registration
 
For registration purposes the file **server_properties.json** should be created and located in the **$CODEINSIGHT_INSTALLDIR/custom_report_scripts/** directory.  This file contains a json with information required to register the report within Code Insight as shown  here:

>     {
>         "core.server.url": "http://localhost:8888" ,
>         "core.server.token" : "Admin authorization token from Code Insight"
>     }

The value for core.server.url is also used within [create_report.py](create_report.py) for any project or inventory based links back to the Code Insight server within a generated report.

If the common **server_properties.json** files is not used then the information the the following files will need to be updated:

[registration.py](registration.py)  -  Update the **baseURL** and **adminAuthToken** values. These settings allow the report itself to be registered on the Code Insight server.

[create_report.py](create_report.py)  -  Update the **baseURL** value. This URL is used for links within the reports.

Report option default values can also be specified in [registration_config.json](registration_config.json).

### Registering the Report

Prior to being able to call the script directly from within Code Insight it must be registered. The [registration.py](registration.py) file can be used to directly register the report once the contents of this repository have been copied into the custom_report_script folder at the base Code Insight installation directory. Default values for the report options can be specified within the [registration.py](registration.py) file


To register this report:

    python registration.py -reg  

To unregister this report:

    python registration.py -unreg

To update this report configuration:

    python registration.py -update

## Usage

This report is executed directly from within Revenera's Code Insight product. From the project reports tab of each Code Insight project it is possible to *generate* the **Claimed File Report** via the Custom Report Framework.

The following report options can be set once the report generation has been initiated:

- Evidence contains strings - a | separated list of strings that can be claimed. i.e. Flexera | Revenera
- Claim files with search term evidence - (True/False) - Consider serch terms when determining to claim files or not
- Mark as reviewed and assigned files to inventory - (True/False)
- Inventory item to assign claimed files to - The name of the inventory file to associated the files with.

The Code Insight Custom Report Framework will provide the following to the custom report when initiated:

- Project ID
- Report ID
- Authorization Token

For this example report these three items are passed on to a batch or sh file which will in turn execute a python script. This script will then:

- Collect data for the report via REST API using the Project ID and Authorization Token
- Take this collected data and generated an html document with details about the project inventory
    - The *"viewable"* file
- Create a zip file of this html document
    - The *"downloadable"* file
- Create a zip file with the viewable file and the downloadable file
- Upload this combined zip file to Code Insight via REST API
- Delete the report artifacts that were created as the script ran

## License

[MIT](LICENSE.TXT)

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.2.1] - 2023-10-16
### Changed
- Update api submodule to prep for tomcat upgrade in 2023R4
- Common requiremtns file for Python 3.6.8

## [1.2.0] - 2023-03-21
### Changed
- Common registration script with registration_config.json file

## [1.1.0] - 2023-02-02
### Changed
- Logo update

## [1.0.7] - 2022-05-23
### Fixed
- Registration updates

## [1.0.6] - 2022-02-10
### Added
- Support for self signed certificates

## [1.0.5] - 2021-12-15
### Changed
- Updated requirements

## [1.0.4] - 2021-12-13
### Added
- Add support for installer
### Changed
- Separate artifact creation
- Common report name and timestamp

## [1.0.3] - 2021-06-07
### Added
- Artifact file name changes based on project ID and time
- RESTAPI submodule update

## [1.0.2] - 2021-06-01
### Added
- Add update capability to registration script
- Add ability to read from common config file for registration

## [1.0.1] - 2021-05-04
### Added
- Initial public release of claimed file report


report_data.py
'''
Copyright 2021 Flexera Software LLC
See LICENSE.TXT for full license text
SPDX-License-Identifier: MIT

Author : sgeary  
Created On : Fri Mar 19 2021
File : report_data.py
'''
import logging

import CodeInsight_RESTAPIs.project.get_project_information
import CodeInsight_RESTAPIs.project.get_project_evidence
import CodeInsight_RESTAPIs.project.get_inventory_summary
import CodeInsight_RESTAPIs.inventory.add_files_to_inventory
import CodeInsight_RESTAPIs.inventory.create_inventory
import CodeInsight_RESTAPIs.inventory.recall_inventory


logger = logging.getLogger(__name__)

#-------------------------------------------------------------------#
def gather_data_for_report(baseURL, projectID, authToken, reportName, reportOptions):
 
    projectInformation = CodeInsight_RESTAPIs.project.get_project_information.get_project_information_summary(baseURL, projectID, authToken)
    projectName = projectInformation["name"]

    fileEvidence = {} # Dict to hold evidience information for each file (claimable or not)

    # Parse report options
    takeAction = reportOptions["takeAction"] 
    stringsToClaim = reportOptions["stringsToClaim"]  
    inventoryItemForClaimedFiles = reportOptions["inventoryItemForClaimedFiles"] 
    isSearchTermClaimable = reportOptions["isSearchTermClaimable"] 

    # See if there are mulitple strings to be claimed and clean them up
    stringsToClaim = stringsToClaim.split("|")
    stringsToClaim = [string.lower().strip() for string in stringsToClaim]

    # Get the evidence gathered
    projectEvidence = CodeInsight_RESTAPIs.project.get_project_evidence.get_project_evidence(baseURL, projectID, authToken)
    for evidence in projectEvidence["data"]:

        fileName = evidence["fileName"]
        filePath = evidence["filePath"]
        remote = evidence["remote"]
        scannedFileId = evidence["scannedFileId"]
        copyrightEvidienceFound = evidence["copyRightMatches"]
        emailUrlEvidenceFound = evidence["emailUrlMatches"]
        licenseEvidenceFound =  evidence["licenseMatches"]
        searchTextMatchEvidenceFound =  evidence["searchTextMatches"]
        exactFileMatchEvidenceFound =  evidence["exactFileMatches"]
        sourceMatchEvidenceFound =  evidence["sourceMatches"]

        fileEvidence[filePath] = {} # Dict to hold the data for each file
        fileEvidence[filePath]["filelink"] =  baseURL + "/codeinsight/FNCI#myprojectdetails/?id=" + str(projectID) + "&tab=workbench&view=file&fileid=" + str(scannedFileId) + "&remote=" + str(remote)

        # See if there was any evidence at all and if so then look further
        if len(copyrightEvidienceFound) or len(emailUrlEvidenceFound) or len(licenseEvidenceFound):
            fileEvidence[filePath]["hasEvidence"] = True
            fileEvidence[filePath]["claimableEvidence"] = {} # Dict to hold the data for each file
            fileEvidence[filePath]["nonclaimableEvidence"] = {} # Dict to hold the data for each file
        
            # For each file see if there is something other than what we are looking for
            # if so this file would need review since there is more too it
            for copyright in copyrightEvidienceFound:
                if any(string in copyright.lower() for string in stringsToClaim):
                    fileEvidence[filePath]["claimableEvidence"]["copyright"] = True
                else:
                    fileEvidence[filePath]["nonclaimableEvidence"]["copyright"] = True

            for emailURL in emailUrlEvidenceFound:
                if any(string in emailURL.lower() for string in stringsToClaim):
                    fileEvidence[filePath]["claimableEvidence"]["emailURL"] = True
                else:
                    fileEvidence[filePath]["nonclaimableEvidence"]["emailURL"] = True
      
            if len(licenseEvidenceFound):
                fileEvidence[filePath]["nonclaimableEvidence"]["license"] = True
            else:
                fileEvidence[filePath]["claimableEvidence"]["license"] = True

            if searchTextMatchEvidenceFound:
                fileEvidence[filePath]["searchTerm"] = True
            else:
                fileEvidence[filePath]["searchTerm"] = False 

            # With an exact match or a source match we want to make sure the file
            # cannot be claimed automatically
            if exactFileMatchEvidenceFound:
                fileEvidence[filePath]["nonclaimableEvidence"]["exactMatch"] = True

            if sourceMatchEvidenceFound:
                fileEvidence[filePath]["nonclaimableEvidence"]["sourceMatch"] = True
                         
        else:
            fileEvidence[filePath]["hasEvidence"] = False


    # Now create seperate dictionaries to hold the data for each catagory
    claimableFiles = {}
    nonclaimableFiles = {}

    # Only worry about the files that contain the evidience to be claimed
    for filePath in fileEvidence:
        if fileEvidence[filePath]["hasEvidence"]:

            # Does the file contain both claimable and nonclaimable evidence?
            if fileEvidence[filePath]["claimableEvidence"] and fileEvidence[filePath]["nonclaimableEvidence"]:
                if fileEvidence[filePath]["searchTerm"]: 
                    fileEvidence[filePath]["nonclaimableEvidence"]["searchTerm"] = True
                nonclaimableFiles[filePath] = fileEvidence[filePath]
            elif fileEvidence[filePath]["claimableEvidence"]:

                # For a search team check to see if the user wants these to be considered or not
                # prior to claiming.  If True then a search term match will not prevent the file
                # from being claimed.  Files need to be compared to see if they can be claimed 
                # already and then check against this criteria

                if isSearchTermClaimable.lower() == "true":
                    if fileEvidence[filePath]["searchTerm"]: 
                        fileEvidence[filePath]["claimableEvidence"]["searchTerm"] = True
                    claimableFiles[filePath] = fileEvidence[filePath]
                else:
                    if fileEvidence[filePath]["searchTerm"]:
                        fileEvidence[filePath]["nonclaimableEvidence"]["searchTerm"] = True
                        nonclaimableFiles[filePath] = fileEvidence[filePath]
                    else:
                        claimableFiles[filePath] = fileEvidence[filePath]


    if takeAction.lower() == "true":
        takeAction = True # To normalize it for passing to report artifact logic

        # Are there any files to claim at all?
        if claimableFiles:
            inventoryID = 0

            # First see if there is already an inventory item to assign the files to
            projectInventorySummary = CodeInsight_RESTAPIs.project.get_inventory_summary.get_all_project_inventory(baseURL, projectID, authToken)
            
            for itemInInventory in projectInventorySummary:            
                # The inventory item to stored claimed files already exists
                if inventoryItemForClaimedFiles == itemInInventory["name"]:
                    inventoryID = itemInInventory["id"]
                    logger.debug("Inventory item %s found with ID of: %s" %(inventoryItemForClaimedFiles, inventoryID))
                    break
                else:
                    pass

            # The item does not exist so create it
            if inventoryID == 0:
                inventoryID = CodeInsight_RESTAPIs.inventory.create_inventory.create_work_in_progress_inventory_item(baseURL, projectID, authToken, inventoryItemForClaimedFiles)
                logger.debug("Work In Progess Inventory item %s created with ID of: %s" %(inventoryItemForClaimedFiles, inventoryID))
                # Recall newly created inventory item
                CodeInsight_RESTAPIs.inventory.recall_inventory.recall_inventory_item(baseURL, inventoryID, authToken)

            try:
                pass
            except:
                logger.error("Inventory item %s does not exist!" %(inventoryItemForClaimedFiles))
                
            # Add the claimable files
            CodeInsight_RESTAPIs.inventory.add_files_to_inventory.add_files_to_inventory_and_mark_as_reviewed(baseURL, inventoryID, authToken, claimableFiles)

        else:
            logger.info("No files to claim")


    else:
        takeAction = False

    
    # Build up the data to return for the report generation
    reportData = {}
    reportData["reportName"] = reportName
    reportData["projectName"] = projectName
    reportData["projectID"] = projectID
    reportData["takeAction"] = takeAction
    reportData["stringsToClaim"] = stringsToClaim
    reportData["inventoryItemForClaimedFiles"] = inventoryItemForClaimedFiles
    reportData["claimableFiles"] = claimableFiles
    reportData["nonclaimableFiles"] = nonclaimableFiles
    
    
    return reportData

create_report.sh
#!/bin/bash
projectId=$1
reportId=$2
authToken=$3
reportOptions=$4

###############################################################################
#  Call the script to collect the data and generate the report
#  This script will create a zip file containing the viewable file
#  combined with another zip file that contains all report artifacts for
#  download.  Since this is executed from the tomcat/bin directory we need to
#  use REPORTDIR to get the location of this shell script since the script is
#  relative to that.
###############################################################################

REPORTDIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"

python3 ${REPORTDIR}/create_report.py -pid $projectId -rid $reportId -authToken $authToken -reportOpts "$reportOptions" 


report_errors.py
'''
Copyright 2021 Flexera Software LLC
See LICENSE.TXT for full license text
SPDX-License-Identifier: MIT

Author : sgeary  
Created On : Mon Apr 12 2021
File : report_errors.py
'''
import logging
import os
import base64

import _version

logger = logging.getLogger(__name__)

#--------------------------------------------------------------------------------#
def create_error_report(reportData):
    logger.info("Entering create_error_report")

    # Dict to hold the complete list of reports
    reports = {}

    htmlFile = generate_error_report(reportData)
    
    reports["viewable"] = htmlFile
    reports["allFormats"] = [htmlFile]

    logger.info("Exiting create_report_artifacts")
    
    return reports 

#------------------------------------------------------------------#
def generate_error_report(reportData):
    logger.info("    Entering generate_error_report")

    reportName = reportData["reportName"]
    reportFileNameBase = reportData["reportFileNameBase"]
    errorMsg = reportData["errorMsg"]
    reportTimeStamp = reportData["reportTimeStamp"] 

    scriptDirectory = os.path.dirname(os.path.realpath(__file__))
    cssFile =  os.path.join(scriptDirectory, "report_branding/css/revenera_common.css")
    logoImageFile =  os.path.join(scriptDirectory, "report_branding/images/logo_reversed.svg")
    iconFile =  os.path.join(scriptDirectory, "report_branding/images/favicon-revenera.ico")
    
    reportName = "Report Creation Failure"
    reportName = reportData["reportName"] + " - Report Creation Failure"
    errorMsg = reportData["errorMsg"]

    #########################################################
    #  Encode the image files
    encodedLogoImage = encodeImage(logoImageFile)
    encodedfaviconImage = encodeImage(iconFile)

    # Grab the current date/time for report date stamp
    htmlFile = reportFileNameBase + ".html"
    
    #---------------------------------------------------------------------------------------------------
    # Create a simple HTML file to display
    #---------------------------------------------------------------------------------------------------
    try:
        html_ptr = open(htmlFile,"w")
    except:
        logger.error("Failed to open htmlfile %s:" %htmlFile)
        raise

    html_ptr.write("<html>\n") 
    html_ptr.write("    <head>\n")
    html_ptr.write("        <!-- Required meta tags --> \n")
    html_ptr.write("        <meta charset='utf-8'>  \n")
    html_ptr.write("        <meta name='viewport' content='width=device-width, initial-scale=1, shrink-to-fit=no'> \n")

    html_ptr.write(''' 
        <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.5.1/css/bootstrap.min.css" integrity="sha384-VCmXjywReHh4PwowAiWNagnWcLhlEJLA5buUprzK8rxFgeH0kww/aWY76TfkUoSX" crossorigin="anonymous">
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/4.1.3/css/bootstrap.css">
    ''')


    html_ptr.write("        <style>\n")

    # Add the contents of the css file to the head block
    try:
        f_ptr = open(cssFile)
        logger.debug("Adding css file details")
        for line in f_ptr:
            html_ptr.write("            %s" %line)
        f_ptr.close()
    except:
        logger.error("Unable to open %s" %cssFile)
        print("Unable to open %s" %cssFile)


    html_ptr.write("        </style>\n")  

    html_ptr.write("    	<link rel='icon' type='image/png' href='data:image/png;base64, {}'>\n".format(encodedfaviconImage.decode('utf-8')))
    html_ptr.write("        <title>%s</title>\n" %(reportName))
    html_ptr.write("    </head>\n") 

    html_ptr.write("<body>\n")
    html_ptr.write("<div class=\"container-fluid\">\n")

    #---------------------------------------------------------------------------------------------------
    # Report Header
    #---------------------------------------------------------------------------------------------------
    html_ptr.write("<!-- BEGIN HEADER -->\n")
    html_ptr.write("<div class='header'>\n")
    html_ptr.write("  <div class='logo'>\n")
    html_ptr.write("    <img src='data:image/svg+xml;base64,{}' style='width: 400px;'>\n".format(encodedLogoImage.decode('utf-8')))
    html_ptr.write("  </div>\n")
    html_ptr.write("<div class='report-title'>%s</div>\n" %reportName)
    html_ptr.write("</div>\n")
    html_ptr.write("<!-- END HEADER -->\n")

    #---------------------------------------------------------------------------------------------------
    # Body of Report
    #---------------------------------------------------------------------------------------------------
    html_ptr.write("<!-- BEGIN BODY -->\n") 
    html_ptr.write("<div class='jumbotron'>\n")
    html_ptr.write("<h1 class='display-4'>Error creating report!</h1>\n")
    for message in errorMsg:
        html_ptr.write("<p class='lead'>%s</p>\n" %message)
    html_ptr.write("</div>\n")

    html_ptr.write("<!-- END BODY -->\n")  
    #---------------------------------------------------------------------------------------------------
    # Report Footer
    #---------------------------------------------------------------------------------------------------
    html_ptr.write("<!-- BEGIN FOOTER -->\n")
    html_ptr.write("<div class='report-footer'>\n")
    html_ptr.write("  <div style='float:right'>Generated on %s</div>\n" %reportTimeStamp)
    html_ptr.write("<br>\n")
    html_ptr.write("  <div style='float:right'>Report Version: %s</div>\n" %_version.__version__)
    html_ptr.write("</div>\n")
    html_ptr.write("<!-- END FOOTER -->\n")  

    html_ptr.write("</div>\n")    

    #---------------------------------------------------------------------------------------------------
    # Add javascript 
    #---------------------------------------------------------------------------------------------------

    html_ptr.write('''

    <script src="https://code.jquery.com/jquery-3.5.1.slim.min.js" integrity="sha384-DfXdz2htPH0lsSSs5nCTpuj/zy4C+OGpamoFVy38MVBnE+IbbVYUew+OrCXaRkfj" crossorigin="anonymous"></script>
    ''')

    html_ptr.write("</body>\n") 
    html_ptr.write("</html>\n") 
    html_ptr.close() 

    logger.info("    Exiting generate_error_report")
    return htmlFile


####################################################################
def encodeImage(imageFile):

    #############################################
    # Create base64 variable for branding image
    try:
        with open(imageFile,"rb") as image:
            encodedImage = base64.b64encode(image.read())
            return encodedImage
    except:
        logger.error("Unable to open %s" %imageFile)
        raise

registration.py
'''
Copyright 2021 Flexera Software LLC
See LICENSE.TXT for full license text
SPDX-License-Identifier: MIT

Author : sgeary  
Created On : Fri Mar 19 2021
File : registration.py
'''

import sys
import os
import stat
import logging
import argparse
import json


import CodeInsight_RESTAPIs.reports.get_reports
import CodeInsight_RESTAPIs.reports.create_report
import CodeInsight_RESTAPIs.reports.delete_report
import CodeInsight_RESTAPIs.reports.update_report

###################################################################################
# Test the version of python to make sure it's at least the version the script
# was tested on, otherwise there could be unexpected results
if sys.version_info <= (3, 5):
    raise Exception("The current version of Python is less than 3.5 which is unsupported.\n Script created/tested against python version 3.8.1. ")
else:
    pass

propertiesFile = "../server_properties.json"  # Created by installer or manually
logfileName = "_custom_report_registration.log"

###################################################################################
#  Set up logging handler to allow for different levels of logging to be capture
logging.basicConfig(format='%(asctime)s,%(msecs)-3d  %(levelname)-8s [%(filename)-25s:%(lineno)-4d]  %(message)s', datefmt='%Y-%m-%d:%H:%M:%S', filename=logfileName, filemode='w',level=logging.DEBUG)
logger = logging.getLogger(__name__)
logging.getLogger("urllib3").setLevel(logging.WARNING)  # Disable logging for requests module

#####################################################################################################
#  Code Insight System Information
#  See if there is a common file for config details
if os.path.exists(propertiesFile):
    try:
        file_ptr = open(propertiesFile, "r")
        configData = json.load(file_ptr)
        baseURL = configData["core.server.url"]
        adminAuthToken = configData["core.server.token"]
        file_ptr.close()
        logger.info("Loading config data from properties file: %s" %propertiesFile)
    except:
        logger.error("Unable to open properties file: %s" %propertiesFile)
else:
    logger.info("Using config data from create_report.py")
    baseURL = "UPDATEME" # i.e. http://localhost:8888 or https://sca.mycodeinsight.com:8443 
    adminAuthToken = "UPDATEME"

#####################################################################################################
# Quick sanity check
if adminAuthToken == "UPDATEME" or baseURL == "UPDATEME":
    logger.error("Make sure baseURL and the admin authorization token have been updated within registration.py")
    print("Make sure baseURL and the admin authorization token have been updated within registration.py")
    sys.exit()


#####################################################################################################
#  Report Details
reportName = "Claimed File Report"  # What is the name to be shown within Code Insight?
enableProjectPickerValue = "false"   # true if a second project can be used within this report
reportOptions = []
#
reportOption = {}
reportOption["name"] = "stringsToClaim"
reportOption["label"] = "Evidence contains strings"
reportOption["description"] = "A string or list of stings separated by '|' that will be claimed if it is the <b>only</b> evidence discovered. For example: <b>foo | bar | foobar</b>"
reportOption["type"] = "string"
reportOption["defaultValue"] = "Flexera | Revenera | Acme"
reportOption["required"] = "true"
reportOption["order"] = "1"
reportOptions.append(reportOption)

reportOption = {}
reportOption["name"] = "isSearchTermClaimable"
reportOption["label"] = "Claim files with search term evidence?  <b>(True/False)</b>"
reportOption["description"] = "<b>True</b> - Claim file if there is search terms evidence in file.<br> <b>False</b> - Do Not claim file if there is search terms evidence in file."
reportOption["type"] = "string"
reportOption["defaultValue"] = "True"
reportOption["required"] = "true"
reportOption["order"] = "2"
reportOptions.append(reportOption)

reportOption = {}
reportOption["name"] = "takeAction"
reportOption["label"] = "Mark as reviewed and assign files to inventory <b>(True/False)</b>"
reportOption["description"] = "<b>True</b> - Mark files as reviewed and assign to inventory.<br> <b>False</b> - Only report on the actions that would be taken if true."
reportOption["type"] = "string"
reportOption["defaultValue"] = "False"
reportOption["required"] = "true"
reportOption["order"] = "3"
reportOptions.append(reportOption)

reportOption = {}
reportOption["name"] = "inventoryItemForClaimedFiles"
reportOption["label"] = "Inventory item to assign claimed files to."
reportOption["description"] = "All files that are claimed will be marked as reviewed and assigned to this inventory item if the above open is <b>True</b>."
reportOption["type"] = "string"
reportOption["defaultValue"] = "My Claimed Files"
reportOption["required"] = "false"
reportOption["order"] = "4"
reportOptions.append(reportOption)


#####################################################################################################
# The path with the custom_report_scripts folder to called via the framework
if sys.platform.startswith('linux'):
    reportHelperScript = "create_report.sh"
elif sys.platform == "win32":
    reportHelperScript = "create_report.bat"
else:
    sys.exit("No script file for operating system")


#####################################################################################################
# Get the directory name in order to register the script
# this will be based on the git repo name is some cases
currentFolderName = os.path.basename(os.getcwd())
reportPath = currentFolderName + "/" + reportHelperScript     

# Create command line argument options
parser = argparse.ArgumentParser()
parser.add_argument('-reg', "--register", action='store_true', help="Register custom reports")
parser.add_argument("-unreg", "--unregister", action='store_true', help="Unegister custom reports")
parser.add_argument("-update", "--update", action='store_true', help="Update a registered custom reports")


#----------------------------------------------------------------------#
def main():
    # See what if any arguments were provided
    args = parser.parse_args()

    if args.register and args.unregister:
        # You can use both options at the same time
        parser.print_help(sys.stderr)
    elif args.register:
        register_custom_reports()
        if sys.platform.startswith('linux'):
            # Make the shell script executable
            os.chmod(reportHelperScript, os.stat(reportHelperScript).st_mode | stat.S_IEXEC)
    elif args.unregister:
        unregister_custom_reports()
    elif args.update:
        update_custom_reports()
    else:
        parser.print_help(sys.stderr)

#-----------------------------------------------------------------------#
def register_custom_reports():
    logger.debug("Entering register_custom_reports")

    # Get the current reports so we can ensure the indexes of the new
    # reports have no conflicts
    try:
        currentReports = CodeInsight_RESTAPIs.reports.get_reports.get_all_currently_registered_reports(baseURL, adminAuthToken)
    except:
        logger.error("Unable to retrieve currently registered reports")
        print("Unable to retrieve currently registered reports.  See log file for details")
        sys.exit()

    # Determine the maximun ID of any current report
    maxReportOrder = max(currentReports, key=lambda x:x['id'])["order"]
    reportOrder = maxReportOrder + 1

    logger.info("Attempting to register %s with a report order of %s" %(reportName, reportOrder))
    print("Attempting to register %s with a report order of %s" %(reportName, reportOrder))

    try:
        reportID = CodeInsight_RESTAPIs.reports.create_report.register_report(reportName, reportPath, reportOrder, enableProjectPickerValue, reportOptions, baseURL, adminAuthToken)
        print("Report registration succeeded! %s has been registered with a report ID of %s" %(reportName, reportID))
        logger.info("Report registration succeeded! %s has been registered with a report ID of %s" %(reportName, reportID))
    except:
        logger.error("Report registration failed! Unable to registered report %s" %reportName)
        print("Report registration failed! Unable to registered report %s.  See log file for details" %reportName)
        sys.exit()


#-----------------------------------------------------------------------#
def unregister_custom_reports():
    logger.debug("Entering unregister_custom_reports")

    try:
        CodeInsight_RESTAPIs.reports.delete_report.unregister_report(baseURL, adminAuthToken, reportName)
        print("%s has been unregistered." %reportName)
        logger.info("%s has been unregistered."%reportName)
    except:
        logger.error("Unable to unregister report %s" %reportName)
        print("Unable to unregister report %s.  See log file for details" %reportName)
        sys.exit()

#-----------------------------------------------------------------------#
def update_custom_reports():
    logger.debug("Entering update_custom_reports")

    try:
        currentReportDetails = CodeInsight_RESTAPIs.reports.get_reports.get_all_currently_registered_reports_by_name(baseURL, adminAuthToken, reportName)
    except:
        logger.error("Unable to retrieve details about report: %s" %reportName)
        print("Unable to retrieve details about report: %s.  See log file for details" %reportName)
        sys.exit()

    reportID = currentReportDetails[0]["id"]
    reportOrder = currentReportDetails[0]["order"]

    logger.info("Attempting to update %s with a report id of %s" %(reportName, reportID))
    print("Attempting to update %s with a report id of %s" %(reportName, reportID))

    try:
        reportID = CodeInsight_RESTAPIs.reports.update_report.update_custom_report(reportName, reportPath, reportID, reportOrder, enableProjectPickerValue, reportOptions, baseURL, adminAuthToken)
        print("%s has been updated" %(reportName))
        logger.info("%s has been updated." %(reportName))
    except:
        logger.error("Unable to update report %s" %reportName)
        print("Unable to update report %s.  See log file for details" %reportName)
        sys.exit() 

#----------------------------------------------------------------------#    
if __name__ == "__main__":
    main()    
